apps/studio/src/hooks/useSearchQuery.ts

```diff
@@ -0,0 +1,62 @@
+import { useCallback, useState } from "react"
+import { useDebounce } from "@uidotdev/usehooks"
+
+import type { SearchResultResource } from "~/server/modules/resource/resource.types"
+import type { ResourceType } from "~prisma/generated/generatedEnums"
+import { trpc } from "~/utils/trpc"
+
+interface UseSearchQueryProps {
+  siteId: string
+  resourceTypes: ResourceType[]
+  onSearchSuccess?: () => void
+}
+export const useSearchQuery = ({
+  siteId,
+  resourceTypes,
+  onSearchSuccess,
+}: UseSearchQueryProps) => {
+  const [searchValue, setSearchValue] = useState("")
+  const debouncedSearchTerm = useDebounce(searchValue, 300)
+
+  const { data, isLoading } = useCallback(() => {
+    return trpc.resource.search.useInfiniteQuery(
+      {
+        siteId,
+        query: debouncedSearchTerm,
+        resourceTypes,
+      },
+      {
+        onSuccess: onSearchSuccess,
+      },
+    )
+  }, [siteId, debouncedSearchTerm, resourceTypes, onSearchSuccess])()
+
+  const resources = useCallback((): SearchResultResource[] => {
+    return data?.pages.flatMap((page) => page.resources) ?? []
+  }, [data])()
+
+  const totalResultsCount = useCallback(() => {
+    return (
+      data?.pages.reduce((acc, page) => acc + (page.totalCount ?? 0), 0) ?? 0
+    )
+  }, [data])()
```

Comment:
this is an antipattern -- if you are using `useCallback` as an IIFE, you should be directly assigning this to a constant (and optionally using `useMemo` instead)

```tsx
const totalResultsCount = useMemo(() => data?.pages.reduce((acc, page) => acc + (page.totalCount ?? 0), 0) ?? 0, [data])
```